### PR TITLE
[1.2.0] Attached SQL methods to the Client [API-1120]

### PR DESCRIPTION
**Pull request by yuce**
This PR:

* Attaches SQL methods to `hazelcast.Client`: `ExecSQL, QuerySQL` and their `SQLOption` variants and `PrepareSQL`.
* Attached SQL methods use an internal SQL connector.
* Added schema query option.
* Ports tests from the SQL driver.
* Ports code sample from the SQL driver.
* Ports documentation from the SQL driver.

See the TDD at: https://hazelcast.atlassian.net/wiki/spaces/HZC/pages/3942187027/Go+Client+SQL+Support+via+Client+API

### CODE REVIEW

File: internal/sql/driver/driver.go [hunk 1]
```diff
@@ -181,49 +176,43 @@ func MakeConfigFromDSN(dsn string) (*client.Config, error) {
...
 		}
-		updateCredentials(&config, u.User)
-		if err := updateConfig(&config, u.Query()); err != nil {
+		updateCredentials(config.Cluster, u.User)
+		if err := updateConfig(config, u.Query()); err != nil {
 			return nil, fmt.Errorf("parsing DSN options: %w", err)
 		}
 	}
 	sc := SerializationConfig()
 	if sc == nil {
```
[the remark this answers is not in the document]
Comment (yuce) [reply internal/sql/driver/driver.go]: Thanks for catching this. Updated and added a test at: https://github.com/hazelcast/hazelcast-go-client/pull/728/commits/541cd773156c9e3ae2ac32e9c2b8c958fe897b8d

File: client.go [hunk 2]
```diff
@@ -360,4 +475,76 @@ func (c *Client) createComponents(config *Config) {
...
+	}
+	tm := t.Milliseconds()
+	// note that the condition below is for t, not tm
+	if t < 0 {
+		tm = -1
+	}
+	s.timeout = &tm
+}
+
+func (s *SQLOptions) validate() error {
```
[the remark this answers is not in the document]
Comment (yuce) [reply client.go:528]: I've changed `SetCursorBufferSize` so it doesn't panic but stores the error in the `SQLConfig` and SQLConfig.validate` returns that error: https://github.com/hazelcast/hazelcast-go-client/pull/728/commits/9ae018d175e10aba7d4dd925046d17aa0b4e9c3b

File: client.go [hunk 3]
```diff
@@ -274,6 +279,116 @@ func (c *Client) RemoveDistributedObjectListener(ctx context.Context, subscripti
...
+	}
+	return c.db.ExecContext(ctx, query, params...)
+}
+
+/*
+QuerySQL runs the given SQL query on the member-side and returns a row iterator.
+This method is used for SQL queries that return rows, such as SELECT and SHOW MAPPINGS.
+Placeholders in the query is replaced by params.
+A placeholder is the question mark (?) character.
+For each placeholder, a corresponding parameter should exist.
```
[the remark this answers is not in the document]
Comment (yuce) [reply client.go:340]: I've changed `param` to `parameter`, etc at: https://github.com/hazelcast/hazelcast-go-client/pull/728/commits/541cd773156c9e3ae2ac32e9c2b8c958fe897b8d

File: client.go [hunk 4]
```diff
@@ -274,6 +279,116 @@ func (c *Client) RemoveDistributedObjectListener(ctx context.Context, subscripti
...
+func (c *Client) QuerySQL(ctx context.Context, query string, params ...interface{}) (*sql.Rows, error) {
+	return c.db.QueryContext(ctx, query, params...)
+}
+
+/*
+QuerySQLWithOptions runs the given SQL query on the member-side and returns a row iterator.
+This method is used for SQL queries that return rows, such as SELECT and SHOW MAPPINGS.
+Placeholders in the query is replaced by params.
+A placeholder is the question mark (?) character.
+For each placeholder, a corresponding parameter should exist.
```
[the remark this answers is not in the document]
Comment (yuce) [reply client.go:365]: I've changed `param` to `parameter`, etc at: https://github.com/hazelcast/hazelcast-go-client/pull/728/commits/541cd773156c9e3ae2ac32e9c2b8c958fe897b8d

**Review by utku-caglayan — APPROVED** [review]
Thanks, looks great!

**Comment by yuce** [comment]
@utku-caglayan 

> Just asking because of curiosity, what is the reason for this change?

Do you mean updating the timeout to `25m`?

**Comment by srknzl** [comment]
@yuce You can see the point Utku commented in the commit. https://github.com/hazelcast/hazelcast-go-client/commit/541cd773156c9e3ae2ac32e9c2b8c958fe897b8d

It is here https://github.com/hazelcast/hazelcast-go-client/commit/541cd773156c9e3ae2ac32e9c2b8c958fe897b8d#diff-b8450a63a6a4040943aba9cf874d490393dfd38dee72823cf80788536a59df71R78

**Comment by srknzl** [comment]
I am going to start reviewing this today, sorry for the delay guys.

**Comment by yuce** [comment]
@srknzl We've talked with @utku-caglayan about it. The tests already took 17-18 minutes before this PR, with the new tests they seem to be takıng over 20 mins.

**Comment by yuce** [comment]
For some reason SQL Driver tests cannot be built, that's the reason of the test failures. I'll check why is that.

File: client.go [hunk 5]
```diff
@@ -360,4 +482,100 @@ func (c *Client) createComponents(config *Config) {
...
+	if t < 0 {
+		tm = -1
+	}
+	s.timeout = &tm
+}
+
+/*
+SetSchema sets the schema name.
+The engine will try to resolve the non-qualified object identifiers from the statement in the given schema.
+If not found, the default search path will be used.
```
Comment (srknzl) [inline client.go:538]:
```suggestion
If not found, the default search path will be used, which looks for objects in the predefined schemas "partitioned" and "public".
```

We may add this part which exist in Java client apidoc.

File: client.go [hunk 6]
```diff
@@ -360,4 +482,100 @@ func (c *Client) createComponents(config *Config) {
...
+If the timeout is reached for a running statement, it will be cancelled forcefully.
+Zero value means no timeout.
+Negative values mean that the value from the server-side config will be used.
+Defaults to -1.
+*/
+func (s *SQLOptions) SetQueryTimeout(t time.Duration) {
+	if t == 0 {
+		s.timeout = nil
+		return
+	}
```
Comment (srknzl) [inline client.go:526]: Giving 0 to timeout is not the same as -1. If a user gives 0, you set timeout to nil, and then in ExtractTimeoutMillis if timeout is nil, you set it to the default timeout value, which is -1.

Isn't this wrong?

File: client.go [hunk 7]
```diff
@@ -360,4 +482,100 @@ func (c *Client) createComponents(config *Config) {
...
+A bigger buffer size may give you a slight performance boost for queries with large result sets at the cost of increased memory consumption.
+Defaults to 4096.
+The given buffer size must be in the non-negative int32 range.
+*/
+func (s *SQLOptions) SetCursorBufferSize(cbs int) {
+	v, err := check.NonNegativeInt32(cbs)
+	if err != nil {
+		s.err = ihzerrors.NewIllegalArgumentError("setting cursor buffer size", err)
+		return
+	}
```
Comment (srknzl) [inline client.go:511]: Is there a specific reason not throwing here but while executing SQL? Note: I have read [this thread](https://github.com/hazelcast/hazelcast-go-client/pull/728/files#r786701195) but could not understand the reason of the change in that comment.

File: client.go [hunk 8]
```diff
@@ -360,4 +482,100 @@ func (c *Client) createComponents(config *Config) {
...
+The engine will try to resolve the non-qualified object identifiers from the statement in the given schema.
+If not found, the default search path will be used.
+The schema name is case-sensitive. For example, foo and Foo are different schemas.
+By default, only the default search path is used.
+*/
+func (s *SQLOptions) SetSchema(schema string) {
+	if schema == "" {
+		s.schema = nil
+		return
+	}
```
Comment (srknzl) [inline client.go:546]: Do we have this logic in Java client? I feel like this is an extra thing. You set schema to nil here if it is empty string, and then you set it to DefaultSchema (which is empty string) in ExtractSchema if it is nil. Why?

File: client.go [hunk 9]
```diff
@@ -360,4 +482,100 @@ func (c *Client) createComponents(config *Config) {
...
+	if s.err != nil {
+		return s.err
+	}
+	if s.cursorBufferSize == 0 {
+		s.cursorBufferSize = idriver.DefaultCursorBufferSize
+	}
+	if s.timeout == nil {
+		v := idriver.DefaultTimeoutMillis
+		s.timeout = &v
+	}
```
Comment (srknzl) [inline client.go:560]: These are not "validating" something, maybe we can move their own method.

**Review by srknzl — COMMENTED** [review]
Just reviewed the client.go file so far. I will continue, but did not want the review to be too big.

File: internal/client/client.go [hunk 10]
```diff
@@ -57,6 +58,34 @@ type Config struct {
 	StatsPeriod   time.Duration
 }
 
+func NewConfig() *Config {
```
Comment (srknzl) [inline internal/client/client.go:61]: Side note: You did not list any config related changes in this PR's description. Are these config related changes intentional? Also can you explain why they are needed.

File: client.go [hunk 11]
```diff
@@ -360,4 +482,100 @@ func (c *Client) createComponents(config *Config) {
 		Logger:               c.ic.Logger,
 	}
 	c.proxyManager = newProxyManager(proxyManagerServiceBundle)
+	c.db = sql.OpenDB(idriver.NewConnectorWithClient(c.ic, true))
```
[the remark this answers is not in the document]
Comment (yuce) [reply client.go:485]: Whenever you create a driver or client, a new internal client is created. I think that's obvious. Sending queries from the same driver or client doesn't create new internal clients.

File: client.go (hunk 7 above)
Comment (yuce) [reply client.go:511]: Throwing a panic is not nice in the public API. We could return an error, but that makes the API harder to use. Also ATM not all setters for SQL options produce errors, but in the future they might.

File: client.go (hunk 9 above)
Comment (yuce) [reply client.go:560]: This function validates that there are no errors and applies the defaults. That's consistent with the rest of the code base.

File: internal/client/client.go (hunk 10 above)
Comment (yuce) [reply internal/client/client.go:61]: There are no config related changes. I've mentioned about the internal client refactoring in the description and this config is only for the internal client. There's no public API change for the config.

File: client.go [hunk 12]
```diff
@@ -274,6 +280,122 @@ func (c *Client) RemoveDistributedObjectListener(ctx context.Context, subscripti
...
+	fmt.Printf("Affected rows: %d\n", cnt)
+
+Note that LastInsertId is not supported and at the moment AffectedRows always returns 0.
+*/
+func (c *Client) ExecSQL(ctx context.Context, query string, params ...interface{}) (driver.Result, error) {
+	return c.db.ExecContext(ctx, query, params...)
+}
+
+/*
+ExecSQLWithOptions runs the given SQL query on the member-side.
```
[the remark this answers is not in the document]
Comment (yuce) [reply client.go:306]: Added a line explaining that at: https://github.com/hazelcast/hazelcast-go-client/pull/728/commits/6d89765f1e4b6dceb68f93e76191d3facb572ae9

File: client.go [hunk 13]
```diff
@@ -274,6 +280,122 @@ func (c *Client) RemoveDistributedObjectListener(ctx context.Context, subscripti
...
+This method is used for SQL queries that don't return rows, such as INSERT, CREATE MAPPING, etc.
+Placeholders in the query is replaced by parameters.
+A placeholder is the question mark (?) character.
+For each placeholder, a corresponding parameter should exist.
```
[the remark this answers is not in the document]
Comment (yuce) [reply client.go]: OK: https://github.com/hazelcast/hazelcast-go-client/pull/728/commits/6d89765f1e4b6dceb68f93e76191d3facb572ae9

File: client.go [hunk 14]
```diff
@@ -274,6 +280,122 @@ func (c *Client) RemoveDistributedObjectListener(ctx context.Context, subscripti
...
+	if err != nil {
+		return nil, err
+	}
+	return c.db.ExecContext(ctx, query, params...)
+}
+
+/*
+QuerySQL runs the given SQL query on the member-side and returns a row iterator.
+This method is used for SQL queries that return rows, such as SELECT and SHOW MAPPINGS.
+Placeholders in the query is replaced by params.
```
[the remark this answers is not in the document]
Comment (yuce) [reply client.go]: Updated at: https://github.com/hazelcast/hazelcast-go-client/pull/728/commits/6d89765f1e4b6dceb68f93e76191d3facb572ae9

File: client.go (hunk 5 above)
Comment (yuce) [reply client.go:536]: Updated at: https://github.com/hazelcast/hazelcast-go-client/pull/728/commits/6d89765f1e4b6dceb68f93e76191d3facb572ae9

File: client.go (hunk 6 above)
Comment (yuce) [reply client.go]: That's right. The idea was enabling the user to "unset" a setting, that doesn't work as expected as you pointed out. Fixed at: https://github.com/hazelcast/hazelcast-go-client/pull/728/commits/6d89765f1e4b6dceb68f93e76191d3facb572ae9

File: client.go (hunk 8 above)
Comment (yuce) [reply client.go]: The idea was enabling the user to "unset" a setting, that doesn't work as expected as you pointed out. Fixed at: [6d89765](https://github.com/hazelcast/hazelcast-go-client/commit/6d89765f1e4b6dceb68f93e76191d3facb572ae9)

File: internal/client/client.go (hunk 10 above)
Comment (yuce) [reply internal/client/client.go:61]: Clarification: The internal client was added at: https://github.com/hazelcast/hazelcast-go-client/pull/695

**Comment by srknzl** [comment]
Changes look good, thanks, continuing reviewing.

File: examples/sql/client-api/employee/main.go [hunk 15]
```diff
@@ -0,0 +1,142 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+This example demonstrates how to use the Hazelcast database/sql driver.
```
Comment (srknzl) [inline examples/sql/client-api/employee/main.go:18]: Is this showing sql driver usage? I think driver usage is like this:

```go
db := sql.Open("DSN")
db.ExecSQL("SELECT..")
```

I bet this is a leftover comment.

File: internal/sql/driver/sql_service.go [hunk 16]
```diff
@@ -49,7 +49,7 @@ type SQLService struct {
 	lg                   *logger.LogAdaptor
 }
 
-func newSQLService(cm *cluster.ConnectionManager, ss *iserialization.Service, fac *cluster.ConnectionInvocationFactory, is *invocation.Service, lg *logger.LogAdaptor) *SQLService {
+func NewSQLService(cm *cluster.ConnectionManager, ss *iserialization.Service, fac *cluster.ConnectionInvocationFactory, is *invocation.Service, lg *logger.LogAdaptor) *SQLService {
```
Comment (srknzl) [inline internal/sql/driver/sql_service.go:52]: We can also leave this newSQLService right? I just want to learn the motivation of the change.

File: internal/sql/driver/driver.go [hunk 17]
```diff
@@ -155,13 +155,10 @@ func (d *Driver) Open(name string) (driver.Conn, error) {
 }
 
 func MakeConfigFromDSN(dsn string) (*client.Config, error) {
-	// TODO: remove hazelcast dependency
-	config := hazelcast.Config{}
-	if err := config.Validate(); err != nil {
-		return nil, err
-	}
```
Comment (srknzl) [inline internal/sql/driver/driver.go:162]: It was strange to validate the config right after creating it, wasn't it? I guess we did not need to do it and therefore removed it now.

File: internal/sql/driver/driver_test.go [hunk 18]
```diff
@@ -53,6 +57,16 @@ func TestParseDSN(t *testing.T) {
...
+		{
+			DSN:           "",
+			Serialization: &serialization.Config{PortableVersion: 2},
+			Pre: func() {
+				idriver.SetSerializationConfig(&serialization.Config{PortableVersion: 2})
+			},
+			Post: func() {
+				idriver.SetSerializationConfig(nil)
+			},
+		},
```
Comment (srknzl) [inline internal/sql/driver/driver_test.go:69]: We don't need to test others right? (Logger and SSL config setter)

File: internal/sql/driver/sql_service.go [hunk 19]
```diff
@@ -202,3 +204,14 @@ func ExtractTimeoutMillis(ctx context.Context) int64 {
...
 }
+
+func ExtractSchema(ctx context.Context) string {
+	if ctx == nil {
+		return DefaultSchema
+	}
+	v := ctx.Value(QuerySchemaKey{})
+	if v == nil {
+		return DefaultSchema
+	}
```
Comment (srknzl) [inline internal/sql/driver/sql_service.go:215]: If this is nil, we should send `null` schema to the cluster, should not we? I mean we should write a null frame to the client message. What will happen in current situation? 

I think empty string is not the same as null string regarding server behaviour.

**Review by srknzl — COMMENTED** [review]
Second part of my review. There is one more pass left for me to review all files.

**Review by srknzl — COMMENTED** [review]
Last part of my review:

File: internal/sql/driver/sql_service.go (hunk 16 above)
Comment (yuce) [reply internal/sql/driver/sql_service.go:52]: Right. Previously I created the SQLService in `hazelcast.Client`, so that function had to be exported, but there's no need to export it now.

File: internal/sql/driver/driver.go (hunk 17 above)
Comment (yuce) [reply internal/sql/driver/driver.go:162]: `Validate` fills in the default values. Previously I've used `hazelcast.Config` there, but I don't use it anymore, so removed that code.

File: internal/sql/driver/sql_service.go (hunk 19 above)
Comment (yuce) [reply internal/sql/driver/sql_service.go:215]: `string` in Go cannot be `nil`. `EncodeNullableForString` adds a null frame when a string is `""`, so this case is correctly handled.

File: serialization/serialization_config.go [hunk 20]
```diff
@@ -38,15 +38,30 @@ type Config struct {
...
-	copy(defs, c.classDefinitions)
-	serializers := map[reflect.Type]Serializer{}
-	for k, v := range c.customSerializers {
-		serializers[k] = v
+	var idFactories []IdentifiedDataSerializableFactory
+	var pFactories []PortableFactory
+	var defs []*ClassDefinition
+	var serializers map[reflect.Type]Serializer
+	if c.identifiedDataSerializableFactories != nil {
+		// this is only necessary to make the cloned value exactly the same when c.identifiedDataSerializableFactories == nil
```
[the remark this answers is not in the document]
Comment (yuce) [reply serialization/serialization_config.go:46]: No. The problem was, a nil slice is equivalent to a slice with 0 length in Go, but `assert.Equal` distinguishes them so makes the tests harder to write. That code only `makes` `idFactories` if `c.identifiedDataSerializableFactories != nil` so `idFactories == nil` if `c.identifiedDataSerializableFactories == nil`

File: sql/driver/driver.go [hunk 21]
```diff
@@ -91,6 +91,20 @@ func WithQueryTimeout(parent context.Context, t time.Duration) context.Context {
 	return context.WithValue(parent, driver.QueryTimeoutKey{}, tm)
 }
 
+/*
+WithSchema returns a copy of parent context which has the given schema name.
+The engine will try to resolve the non-qualified object identifiers from the statement in the given schema.
+If not found, the default search path will be used.
```
[the remark this answers is not in the document]
Comment (yuce) [reply sql/driver/driver.go:97]: They seem to be equivalent to me. What's different with them (except the context stuff)?

File: sql_it_test.go [hunk 22]
```diff
@@ -0,0 +1,581 @@
...
+		rec := NewRecordWithDateTime(&dt)
+		_, err := client.ExecSQL(context.Background(), fmt.Sprintf(`INSERT INTO "%s" (__key, datevalue, timevalue, timestampvalue, timestampwithtimezonevalue) VALUES(?, ?, ?, ?, ?)`, mapName),
+			1, *rec.DateValue, *rec.TimeValue, *rec.TimestampValue, *rec.TimestampWithTimezoneValue)
+		if err != nil {
+			t.Fatal(err)
+		}
+		targetDate := time.Date(2021, 12, 22, 0, 0, 0, 0, time.Local)
+		targetTime := time.Date(0, 1, 1, 23, 40, 12, 3400, time.Local)
+		targetTimestamp := time.Date(2021, 12, 22, 23, 40, 12, 3400, time.Local)
+		targetTimestampWithTimezone := time.Date(2021, 12, 22, 23, 40, 12, 3400, time.FixedZone("", -5*60*60))
```
[the remark this answers is not in the document]
Comment (yuce) [reply sql_it_test.go:356]: Could you clarify your point?

File: sql_it_test.go [hunk 23]
```diff
@@ -0,0 +1,581 @@
...
+		target := serialization.JSON(`{"name":"Ford Prefect"}`)
+		assert.Equal(t, target, v)
+	})
+}
+
+func testSQLQuery(t *testing.T, keyFmt, valueFmt string, keyFn, valueFn func(i int) interface{}, opts *hz.SQLOptions) {
+	fn := func(config *hz.Config) {
+		config.Serialization.SetGlobalSerializer(&it.PanicingGlobalSerializer{})
+	}
+	it.SQLTesterWithConfigBuilder(t, fn, func(t *testing.T, client *hz.Client, config *hz.Config, m *hz.Map, mapName string) {
```
[the remark this answers is not in the document]
Comment (yuce) [reply sql_it_test.go:493]: We use the `PanicingGlobalSerializer` to know when the serialization falls back to the default serializer.

File: doc.go [hunk 24]
```diff
@@ -174,6 +174,44 @@ If you don't want to receive any distributed object events, use client.RemoveDis
...
+	- int8 (tinyint)
+	- int16 (smallint)
+	- int32 (integer)
+	- int64 (bigint)
+	- bool (boolean)
+	- float32 (real)
+	- float64 (double)
+	- types.Decimal (decimal)
+	- time.Time (date) Detected by checking: hour == minute == second == nanoseconds = 0
+	- time.Time (time) Detected by checking: year == 0, month == day == 1
```
[the remark this answers is not in the document]
Comment (yuce) [reply doc.go:211]: I've changed the code to match the doc at: https://github.com/hazelcast/hazelcast-go-client/pull/728/commits/858ee6604d144eb445053d128ef8764ce720e156

File: doc.go [hunk 25]
```diff
@@ -174,6 +174,44 @@ If you don't want to receive any distributed object events, use client.RemoveDis
...
+	- time.Time (timestamp) Detected by checking: hour == minute == second == nanoseconds = 0, timezone == time.Local
```
[the remark this answers is not in the document]
Comment (yuce) [reply doc.go:212]: Updated at: https://github.com/hazelcast/hazelcast-go-client/pull/728/commits/858ee6604d144eb445053d128ef8764ce720e156

File: doc.go [hunk 26]
```diff
@@ -174,6 +174,44 @@ If you don't want to receive any distributed object events, use client.RemoveDis
...
+	- time.Time (timestamp with time zone) Detected by checking: hour == minute == second == nanoseconds = 0, timezone != time.Local
```
[the remark this answers is not in the document]
Comment (yuce) [reply doc.go:213]: Updated at: https://github.com/hazelcast/hazelcast-go-client/pull/728/commits/858ee6604d144eb445053d128ef8764ce720e156

File: examples/sql/client-api/advanced_query_options/main.go [hunk 27]
```diff
@@ -0,0 +1,72 @@
...
+
+func main() {
+	ctx := context.Background()
+	client, err := hazelcast.StartNewClient(ctx)
+	if err != nil {
+		panic(fmt.Errorf("creating the client"))
+	}
+	defer client.Shutdown(ctx)
+	opts := hazelcast.SQLOptions{}
+	opts.SetSchema("parititioned")
```
[the remark this answers is not in the document]
Comment (yuce) [reply examples/sql/client-api/advanced_query_options/main.go]: No, it wasn't. Fixed at: https://github.com/hazelcast/hazelcast-go-client/pull/728/commits/858ee6604d144eb445053d128ef8764ce720e156

File: examples/sql/client-api/employee/main.go (hunk 15 above)
Comment (yuce) [reply examples/sql/client-api/employee/main.go]: That's right, copy/paste hazard. Removed at: https://github.com/hazelcast/hazelcast-go-client/pull/728/commits/858ee6604d144eb445053d128ef8764ce720e156

File: internal/sql/driver/sql_service.go (hunk 16 above)
Comment (yuce) [reply internal/sql/driver/sql_service.go:52]: Since it's internal API, I left it as is.

File: internal/sql/driver/driver_test.go (hunk 18 above)
Comment (yuce) [reply internal/sql/driver/driver_test.go]: I've added them at: https://github.com/hazelcast/hazelcast-go-client/pull/728/commits/858ee6604d144eb445053d128ef8764ce720e156

File: sql_test.go [hunk 28]
```diff
@@ -0,0 +1,116 @@
...
+	"time"
+
+	"github.com/stretchr/testify/assert"
+
+	"github.com/hazelcast/hazelcast-go-client/internal/it/runtime"
+)
+
+type cursorBufferSizeTestCase struct {
+	V         int
+	T         *int32
```
[the remark this answers is not in the document]
Comment (yuce) [reply sql_test.go]: OK, changed at: https://github.com/hazelcast/hazelcast-go-client/pull/728/commits/858ee6604d144eb445053d128ef8764ce720e156

File: sql_it_test.go [hunk 29]
```diff
@@ -0,0 +1,581 @@
...
+			i++
+		}
+		if valueFmt == "decimal" {
+			// assert.Equal does not seem to work for *big.Int values.
+			// Since types.Decimal includes a *big.Int, it doesn't work for it either.
+			// Hence the special treatment for decimal.
+			assert.Equal(t, len(target), len(entries))
+			for i := 0; i < len(target); i++ {
+				t.Run(fmt.Sprintf("decimal-%d", i), func(t *testing.T) {
+					bt := target[i].Value.(types.Decimal)
```
[the remark this answers is not in the document]
Comment (yuce) [reply sql_it_test.go:519]: OK, at: https://github.com/hazelcast/hazelcast-go-client/pull/728/commits/858ee6604d144eb445053d128ef8764ce720e156

File: sql_it_test.go [hunk 30]
```diff
@@ -0,0 +1,581 @@
...
+
+const (
+	factoryID                 = 100
+	recordClassID             = 1
+	recordWithDateTimeClassID = 2
+)
+
+type Record struct {
+	DecimalValue  *types.Decimal
+	NullValue     interface{}
```
[the remark this answers is not in the document]
Comment (yuce) [reply sql_it_test.go]: OK, removed at: https://github.com/hazelcast/hazelcast-go-client/pull/728/commits/858ee6604d144eb445053d128ef8764ce720e156

File: sql_it_test.go [hunk 31]
```diff
@@ -0,0 +1,581 @@
...
+			VarcharValue:  "hello",
+			TinyIntValue:  -128,
+			SmallIntValue: 32767,
+			IntegerValue:  -27,
+			BigIntValue:   38,
+			BoolValue:     true,
+			RealValue:     -5.32,
+			DoubleValue:   12.789,
+			DecimalValue:  &dec,
+		}}
```
[the remark this answers is not in the document]
Comment (yuce) [reply sql_it_test.go]: Yes, changed at: https://github.com/hazelcast/hazelcast-go-client/pull/728/commits/858ee6604d144eb445053d128ef8764ce720e156

File: doc.go (hunk 26 above)
[the remark this answers is not in the document]
Comment (srknzl) [reply doc.go:213]: If a user passes a zero value `time.Time` what happens? According to the code we use `javaLocalTimeSerializer` but we don't document that. If you agree that we should document this feel free to change documentation. If not, you can ignore my comment.

[the remark this answers is not in the document]
Comment (srknzl) [reply doc.go:213]: Another option is to document the algorithm in plain English. It may be simpler that way

File: sql/driver/driver.go (hunk 21 above)
[the remark this answers is not in the document]
Comment (srknzl) [reply sql/driver/driver.go:97]: The part `, which looks for objects in the predefined schemas "partitioned" and "public".` does not exist in WithSchema.

In cursor buffer size methods, in WithCursorBufferSize place we say `Panics if parent context is nil, or given buffer size is not in the positive int32 range.` in the other we say `The given buffer size must be in the non-negative int32 range.`. I think the one in WithCursorBufferSize is wrong

File: sql_it_test.go (hunk 22 above)
[the remark this answers is not in the document]
Comment (srknzl) [reply sql_it_test.go:345]: I am talking about the string in time.FixedZone. We pass "A/B" while creating the record but we assert it to be empty string. I was just wondering why. Also what does A/B mean?

File: sql_it_test.go (hunk 23 above)
[the remark this answers is not in the document]
Comment (srknzl) [reply sql_it_test.go:482]: I think it.SQLTester does that too https://github.com/hazelcast/hazelcast-go-client/blob/e95f792c226cc86690a6de0c82ea5748b5737a31/internal/it/sql.go#L34

File: sql_it_test.go (hunk 22 above)
[the remark this answers is not in the document]
Comment (yuce) [reply sql_it_test.go:345]: A/B has no meaning. The codec for the timestamps does not support location names, so that information cannot be restored, hence it is empty when decoded.

File: sql/driver/driver.go (hunk 21 above)
[the remark this answers is not in the document]
Comment (yuce) [reply sql/driver/driver.go:97]: Oops, I missed those. Updated at: https://github.com/hazelcast/hazelcast-go-client/pull/728/commits/858ee6604d144eb445053d128ef8764ce720e156

**Review by srknzl — APPROVED** [review]
LGTM Thanks a lot

**Comment by yuce** [comment]
@srknzl @utku-caglayan Thanks a lot for the reviews!
